aptos-move/framework/aptos-framework/sources/transaction_validation.move

```diff
@@ -270,15 +270,27 @@ module aptos_framework::transaction_validation {
             error::out_of_range(PROLOGUE_ECANT_PAY_GAS_DEPOSIT),
         );
 
+        let amount_to_burn = 0;
         if (features::collect_and_distribute_gas_fees()) {
+            // TODO(gas): We might want to distinguish the refundable part of the charge and burn it or track
+            // it separately, so that we don't increase the total supply by refunding.
+
             // If transaction fees are redistributed to validators, collect them here for
             // later redistribution.
             transaction_fee::collect_fee(gas_payer, transaction_fee_amount);
         } else {
             // Otherwise, just burn the fee.
             // TODO: this branch should be removed completely when transaction fee collection
             // is tested and is fully proven to work well.
-            transaction_fee::burn_fee(gas_payer, transaction_fee_amount);
+            amount_to_burn = transaction_fee_amount;
+        };
+
+        if (amount_to_burn > storage_fee_refunded) {
```

Comment:
Why not just mint first. Then charge gas?

Reply:
One less function call and the logical is clearer (to me). Lmk you feel strongly about it.

Reply:
Not strongly so this is fine :)

Reply:
this is ugly to me with the fee distribution branching above